### PR TITLE
PG16: Prohibit use of multi-node

### PR DESCRIPTION
**Pull request by fabriziomello**
Since multi-node is not supported on PG16, add errors to multi-node functions when run on this PostgreSQL version. 

Disable-check: force-changelog-file

### CODE REVIEW

**Review by svenklemm — APPROVED** [review]

**Comment by maxhertrampf** [comment]
> Since multi-node has being deprecated

Sorry, what now? TimescaleDB is dropping multi node support? Is there any more info on that? 😮

File: src/hypertable.c [hunk 1]
```diff
@@ -2046,6 +2047,13 @@ ts_hypertable_create(PG_FUNCTION_ARGS)
 Datum
 ts_hypertable_distributed_create(PG_FUNCTION_ARGS)
 {
+#if PG16_GE
+	ereport(ERROR,
+			(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+			 errmsg("distributed hypertable is not supported"),
+			 errhint("Multi-node is not supported anymore on PostgreSQL >= 16."),
+			 errdetail("Use the \"create_hypertable\" to create a regular hypertable instead.")));
+#endif
```
Comment (mkindahl) [inline src/hypertable.c:2056]: Isn't this effectively an API-breaking change, which we should not introduce in a minor release? Features can be deprecated in minor releases, but removing support for them does not sound like a good idea. Also, we deprecate it for all PG versions, not only PG16.

```suggestion
	ereport(WARNING,
			(errcode(ERRCODE_WARNING_DEPRECATED_FEATURE),
			 errmsg("distributed hypertables are deprecated"),
			 errhint("Distributed hypertables are deprecated and will be removed in the future.")));
```

**Review by mkindahl — CHANGES_REQUESTED** [review]
IMHO, we should not remove features in minor releases and this is effectively an API-breaking change. We should deprecate the feature and the actual removal (and API-breaking change) is sometimes in the future.

File: src/hypertable.c [hunk 2]
```diff
@@ -2046,6 +2047,13 @@ ts_hypertable_create(PG_FUNCTION_ARGS)
 Datum
 ts_hypertable_distributed_create(PG_FUNCTION_ARGS)
 {
+#if PG16_GE
+	ereport(ERROR,
+			(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+			 errmsg("distributed hypertable is not supported"),
+			 errhint("Multi-node is not supported anymore on PostgreSQL >= 16."),
+			 errdetail("Use the \"create_hypertable\" to create a regular hypertable instead.")));
```
Comment (erimatnor) [inline src/hypertable.c:2055]:
```suggestion
	ereport(ERROR,
			(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
			 errmsg("distributed hypertables are not supported on PostgreSQL 16 and greater"),
			 errhint("Use \"create_hypertable\" to create a regular hypertable instead.")));
```

File: tsl/src/data_node.c [hunk 3]
```diff
@@ -857,18 +857,37 @@ data_node_add_internal(PG_FUNCTION_ARGS, bool set_distid)
 Datum
 data_node_add(PG_FUNCTION_ARGS)
 {
+#if PG16_GE
+	ereport(ERROR,
+			(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+			 errmsg("adding data node is not supported"),
+			 errhint("Multi-node is not supported anymore on PostgreSQL >= 16.")));
```
Comment (erimatnor) [inline tsl/src/data_node.c:864]: Again, hints should give an idea of how you can "fix" the problem, so suggest turning the hint into details in this case:
```suggestion
	ereport(ERROR,
			(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
			 errmsg("adding data node is not supported"),
			 errdetail("Multi-node is not supported anymore on PostgreSQL >= 16.")));
```

**Comment by erimatnor** [comment]
> > Since multi-node has being deprecated
> 
> Sorry, what now? TimescaleDB is dropping multi node support? Is there any more info on that? 😮

To speed up the release of support for PG16, we are not going to support multi-node when run on PG16.

File: src/hypertable.c (hunk 2 above)
Comment (fabriziomello) [reply src/hypertable.c:2055]: Ooops... fixed!

File: tsl/src/data_node.c (hunk 3 above)
Comment (fabriziomello) [reply tsl/src/data_node.c:864]: Ooops... fixed!

File: src/hypertable.c (hunk 1 above)
Comment (fabriziomello) [reply src/hypertable.c]: Fixed!

**Review by mkindahl — APPROVED** [review]

File: src/hypertable.c [hunk 4]
```diff
@@ -2046,6 +2047,18 @@ ts_hypertable_create(PG_FUNCTION_ARGS)
...
+			(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+			 errmsg("distributed hypertable is not supported"),
+			 errdetail("Multi-node is not supported anymore on PostgreSQL >= 16."),
+			 errhint("Use the \"create_hypertable\" to create a regular hypertable instead.")));
+#else
+	ereport(WARNING,
+			(errcode(ERRCODE_WARNING_DEPRECATED_FEATURE),
+			 errmsg("distributed hypertable is deprecated"),
+			 errdetail("Multi-node is deprecated and will be removed in future releases.")));
+#endif
```
[the remark this answers is not in the document]
Comment (fabriziomello) [reply src/hypertable.c]: Fixed!

**Pull request merged**